xds/src/main/java/io/grpc/xds/EnvoyProtoData.java

```diff
@@ -386,55 +471,98 @@ public String toString() {
           .toString();
     }
 
-    static Route fromEnvoyProtoRoute(io.envoyproxy.envoy.api.v2.route.Route proto) {
-      RouteMatch routeMatch = RouteMatch.fromEnvoyProtoRouteMatch(proto.getMatch());
-      RouteAction routeAction = null;
-      if (proto.hasRoute()) {
-        routeAction = RouteAction.fromEnvoyProtoRouteAction(proto.getRoute());
+    @Nullable
+    static StructOrError<Route> fromEnvoyProtoRoute(io.envoyproxy.envoy.api.v2.route.Route proto) {
+      StructOrError<RouteMatch> routeMatch = RouteMatch.fromEnvoyProtoRouteMatch(proto.getMatch());
+      if (routeMatch == null) {
+        return null;
+      }
+      if (routeMatch.getErrorDetail() != null) {
+        return StructOrError.fromError(
+            "Invalid route [" + proto.getName() + "] : " + routeMatch.getErrorDetail());
+      }
+
+      StructOrError<RouteAction> routeAction;
+      switch (proto.getActionCase()) {
+        case ROUTE:
+          routeAction = RouteAction.fromEnvoyProtoRouteAction(proto.getRoute());
+          break;
+        case REDIRECT:
+          return StructOrError.fromError("Unsupported action type: redirect");
+        case DIRECT_RESPONSE:
+          return StructOrError.fromError("Unsupported action type: direct_response");
+        case FILTER_ACTION:
+          return StructOrError.fromError("Unsupported action type: filter_action");
+        case ACTION_NOT_SET:
+          // fall through
+        default:
+          return StructOrError.fromError("Unknown action type: " + proto.getActionCase());
       }
-      return new Route(routeMatch, routeAction);
+      if (routeAction.getErrorDetail() != null) {
+        return StructOrError.fromError(
+            "Invalid route [" + proto.getName() + "] : " + routeAction.getErrorDetail());
+      }
+      return StructOrError.fromStruct(new Route(routeMatch.getStruct(), routeAction.getStruct()));
     }
   }
 
   /** See corresponding Envoy proto message {@link io.envoyproxy.envoy.api.v2.route.RouteMatch}. */
   static final class RouteMatch {
-    private final String prefix;
```

Comment:
Justification: use `null` for unset values as in proto only one of prefix, path, regex can be set. It is ambiguous for converted objects to use `""` for unset fields given that `""` itself can be a set value.

Reply:
"" is ambiguous anyway. If someone sets a nonempty value and then sets it to empty value, they probably mean to unset it.

Reply:
That's why I changed to use `null` if the corresponding filed in the proto is unset.

Reply:
Sorry I replied to the wrong thread, I thought this was a comment about the environment variable in another PR

Reply:
Ok, for that. You do not need to consider that much. Out implementation should be based on normal usage and is not intended to be failsafe in terms of undocumented usage.